### DISCUSSION TITLE
Memory issues with new datasets after 20-30 epochs

### DISCUSSION THREAD

**Discussion by shenberg** [coqui-ai/TTS#2366] [General Q&A]
Hiya,

I've been training Overflow and HifiGAN on new datasets and getting a bunch of issues that all come down to memory running out after many epochs. The version I'm using was forked off of the `dev` branch on Feb 7, 2023, so the last commit it has is c7184dcef95a421711108e057aecc451eca60b78 ("Linter fix").

The platform is a 32-core AMD EPYC machine with a Titan RTX (24GB) running Ubuntu 20.04, CUDA 11.4, Python 3.10, torch==1.13.1

In order to get Overflow to train on my dataset, I had to reduce the batch size to 10: the data is around 3,000 utterances, and without shuffling it overfit early, and the last batch was also behaving poorly. Shuffling required decreasing batch size.

This is my first unpleasant surprise: I got a batch size of 12 running successfully with `shuffle=False` and `start_by_longest=True` in the config, and I verified that the sample with the longest transcription also has the longest audio, so I know that in terms of memory size, the first batch with these settings is the largest possible memory allocation.

And yet, if I run training with shuffle=True, drop_last=True, batch_size=12, I get an OoM crash after a single-digit number of epochs (a few hundred training steps). Reducing batch size to 10 gets me to high dual-digit epoch count (10k-30k steps) until I crash.

The crash is a CUDA OoM trying to allocate a small allocation. The main interesting part is that PyTorch reports ~20GB allocated, ~23GB reserved, and fails to allocate ~14MB. This implies heap fragmentation?

I'm getting similar behavior for my HifiGAN run, but I have spent much less time trying to make it run so I won't expand on it for now.

Note: I tried to run this training procedure on a bigger machine: 48-core machine with A100 40GB (Ubuntu 22.04, CUDA 11.6, the rest is the same), the result was crashes at roughly similar times, only this time the crash was a std::terminate() exception.

I'm at a loss as to how to proceed: I'm getting functional models but it's a pain to work around consistent crashes. Any ideas?
